lib/IDE/TypeReconstruction.cpp

```diff
@@ -866,14 +866,27 @@ static void VisitNodeAssociatedTypeRef(
       ArchetypeType *archetype = type->getAs<ArchetypeType>();
       if (archetype) {
         Identifier identifier = ast->getIdentifier(ident->getText());
-        if (archetype->hasNestedType(identifier)) {
-          Type nested = archetype->getNestedTypeValue(identifier);
-          if (nested) {
-            result._types.push_back(nested);
-            result._module = type_result._module;
-            return;
+        Type nested;
+        if (archetype->hasNestedType(identifier))
+          nested = archetype->getNestedTypeValue(identifier);
+        else if (ProtocolDecl *self_protocol = archetype->getSelfProtocol()) {
```

Comment:
Discussed in person. What Enrico wants to do still requires this hack, but a patch I have in another PR cleans this up (basically by nuking the SelfTypeRef mangling altogether). Go ahead and merge this PR if you want.